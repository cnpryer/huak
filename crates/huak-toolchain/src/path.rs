use crate::Error;
use std::path::Path;

pub(crate) fn name_from_path(path: &Path) -> Result<&str, Error> {
    let Some(name) = path
        .components()
        .last()
        .map(std::path::Component::as_os_str)
        .and_then(|name| name.to_str())
    else {
        return Err(Error::InvalidToolchain(
            "could not parse name from path".to_string(),
        ));
    };

    Ok(name)
}
