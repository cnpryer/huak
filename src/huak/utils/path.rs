use std::path::{Path, PathBuf};

use anyhow::{self, Error};
use fs_extra::dir;

/// Return the filename from a `Path`.
pub fn parse_filename(path: &Path) -> Result<&str, Error> {
    // Attempt to convert OsStr to str.
    let name = match path.file_name() {
        Some(f) => f.to_str(),
        _ => return Err(anyhow::format_err!("failed to read name from path")),
    };

    // If a str was failed to be parsed error.
    if name.is_none() {
        return Err(anyhow::format_err!(
            "failed to convert filename from {} to string",
            path.display()
        ));
    }

    Ok(name.unwrap())
}

/// Convert a `Path` to a &str.
pub fn to_string(path: &Path) -> Result<&str, anyhow::Error> {
    let pip_path = match path.to_str() {
        Some(s) => s,
        None => {
            return Err(anyhow::format_err!(
                "failed to convert {} to a string",
                path.display()
            ))
        }
    };

    Ok(pip_path)
}

/// Search for manifest files using a path `from` to start from and
/// `steps` to recurse.
pub fn search_parents_for_filepath(
    from: &Path,
    filename: &str,
    steps: usize,
) -> Result<Option<PathBuf>, anyhow::Error> {
    if steps == 0 {
        return Ok(None);
    }

    if from.join(filename).exists() {
        return Ok(Some(from.join(filename)));
    }

    if let Some(parent) = from.parent() {
        return search_parents_for_filepath(parent, filename, steps - 1);
    }

    Ok(None)
}

/// Copies one directory into another.
pub fn copy_dir(from: &PathBuf, to: &PathBuf) -> bool {
    if !Path::new(from).is_file() {
        eprintln!("resource archive does not exist");
    }

    if !Path::new(to).is_dir() {
        eprintln!("`to` {} does not exist", to.display());
    }

    // Copy mock project dir to target dir
    let copy_options = dir::CopyOptions::new();
    dir::copy(from.as_path(), to.as_path(), &copy_options).unwrap();

    true
}

#[cfg(test)]
mod tests {
    use tempfile::tempdir;

    use super::*;

    use crate::utils::test_utils::get_resource_dir;

    #[test]
    fn test_copy_dir() {
        let tmp = tempdir().unwrap().into_path().to_path_buf();
        let from = get_resource_dir().join("mock-project");

        copy_dir(&from, &tmp);

        assert!(tmp.join("mock-project").exists());
        assert!(tmp.join("mock-project").join("pyproject.toml").exists());
    }

    #[test]
    fn test_search_parents_for_filepath() {
        let tmp = tempdir().unwrap().into_path().to_path_buf();
        let from = get_resource_dir().join("mock-project");

        copy_dir(&from, &tmp);

        let res = search_parents_for_filepath(
            &tmp.join("mock-project").join("src"),
            "pyproject.toml",
            5,
        );

        assert!(res.unwrap().unwrap().exists());
    }
}
