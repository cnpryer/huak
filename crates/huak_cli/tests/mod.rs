use std::path::PathBuf;

/// The resource directory found in the Huak repo used for testing purposes.
fn test_resources_dir_path() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .parent()
        .unwrap()
        .join("dev-resources")
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta_cmd::assert_cmd_snapshot;
    use std::process::Command;

    #[test]
    fn test_activate_help() {
        assert_cmd_snapshot!(Command::new("huak")
            .arg("activate")
            .arg("--help"));
    }

    #[test]
    fn test_add_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("add").arg("--help"));
    }

    #[test]
    fn test_build_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("build").arg("--help"));
    }

    #[test]
    fn test_clean_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("clean").arg("--help"));
    }

    #[test]
    fn test_completion_help() {
        assert_cmd_snapshot!(Command::new("huak")
            .arg("completion")
            .arg("--help"));
    }

    #[test]
    fn test_fix_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("fix").arg("--help"));
    }

    #[test]
    fn test_fmt_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("fmt").arg("--help"));
    }

    #[test]
    fn test_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("help"));
        assert_cmd_snapshot!(Command::new("huak").arg("--help"));
    }

    #[test]
    fn test_init_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("init").arg("--help"));
    }

    #[test]
    fn test_install_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("install").arg("--help"));
    }

    #[test]
    fn test_lint_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("lint").arg("--help"));
    }

    #[test]
    fn test_new_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("new").arg("--help"));
    }

    #[test]
    fn test_publish_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("publish").arg("--help"));
    }

    #[test]
    fn test_python_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("python").arg("--help"));
    }

    #[test]
    fn test_remove_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("remove").arg("--help"));
    }

    #[test]
    fn test_run_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("run").arg("--help"));
    }

    #[test]
    fn test_test_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("test").arg("--help"));
    }

    #[test]
    fn test_update_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("update").arg("--help"));
    }

    #[test]
    fn test_version_help() {
        assert_cmd_snapshot!(Command::new("huak").arg("version").arg("--help"));
    }

    #[test]
    fn test_version() {
        let from = test_resources_dir_path().join("mock-project");
        assert_cmd_snapshot!(Command::new("huak")
            .arg("version")
            .arg("--no-color")
            .current_dir(from))
    }
}
