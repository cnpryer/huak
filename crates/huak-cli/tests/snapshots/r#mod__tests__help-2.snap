---
source: crates/huak_cli/tests/mod.rs
info:
  program: huak
  args:
    - "--help"
---
success: true
exit_code: 0
----- stdout -----
A Python package manager written in Rust and inspired by Cargo.

Usage: huak [OPTIONS] <COMMAND>

Commands:
  activate    Activate the virtual environment
  add         Add dependencies to the project
  build       Build tarball and wheel for the project
  clean       Remove tarball and wheel from the built project
  completion  Generates a shell completion script for supported shells
  fix         Auto-fix fixable lint conflicts
  fmt         Format the project's Python code
  init        Initialize the existing project
  install     Install the dependencies of an existing project
  lint        Lint the project's Python code
  new         Create a new project at <path>
  publish     Builds and uploads current project to a registry
  python      Manage Python installations
  remove      Remove dependencies from the project
  run         Run a command within the project's environment context
  test        Test the project's Python code
  update      Update the project's dependencies
  version     Display the version of the project
  help        Print this message or the help of the given subcommand(s)

Options:
  -q, --quiet     
      --no-color  
  -h, --help      Print help
  -V, --version   Print version

----- stderr -----

